napari/plugins/_hookexec.py

```diff
@@ -0,0 +1,110 @@
+tracebacks and feedback if one of them errors.
+
+The code here was submitted to pluggy as a PR:
+https://github.com/pytest-dev/pluggy/pull/253
+
+
+Example
+-------
+Instead of calling hooks the normal way:
+>>> plugin_manager.hook.my_hook(arg=arg)
+
+you can use _hookexec from this module as follows:
+>>> (result, imp) = _hookexec(
+...                     plugin_manager.hook.my_hook,
+...                     arg=arg,
+...                     with_impl=True,
+...                 )
+"""
+import sys
+from typing import Optional, Sequence, Any
+
+from pluggy.callers import HookCallError, _raise_wrapfail, _Result
+from pluggy.hooks import HookImpl
+
+
+def _multicall(hook_impls, caller_kwargs, firstresult=False):
+    """Execute a call into multiple python functions/methods and return the
+    result(s).
+
+    ``caller_kwargs`` comes from _HookCaller.__call__().
+    If ``caller_kwargs`` contains a key ``with_impl`` that evaluates to true,
+    results will be returned as 2-tuples of (result, hook_impl) instead of the
+    bare result.
+    """
+    __tracebackhide__ = True
+    results = []
+    impl_hits = []
+    excinfo = None
+    with_impl = caller_kwargs.pop("with_impl", False)
+    try:  # run impl and wrapper setup functions in a loop
+        teardowns = []
+        try:
+            for hook_impl in reversed(hook_impls):
+                try:
+                    args = [
+                        caller_kwargs[argname]
+                        for argname in hook_impl.argnames
+                    ]
+                except KeyError:
+                    for argname in hook_impl.argnames:
+                        if argname not in caller_kwargs:
+                            raise HookCallError(
+                                "hook call must provide argument %r"
+                                % (argname,)
+                            )
+
+                if hook_impl.hookwrapper:
+                    try:
+                        gen = hook_impl.function(*args)
+                        next(gen)  # first yield
+                        teardowns.append(gen)
+                    except StopIteration:
+                        _raise_wrapfail(gen, "did not yield")
+                else:
+                    res = hook_impl.function(*args)
+                    if res is not None:
+                        results.append(res)
+                        if with_impl:
+                            impl_hits.append(hook_impl)
+                        if firstresult:  # halt further impl calls
+                            break
+        except BaseException:
+            excinfo = sys.exc_info()
+    finally:
+        if firstresult:  # first result hooks return a single value
+            outcome = _Result(results[0] if results else None, excinfo)
+        else:
+            outcome = _Result(results, excinfo)
+
+        # run all wrapper post-yield blocks
+        for gen in reversed(teardowns):
```

Comment:
Why do you reverse this list? And if so, why not make teardowns a stack instead of a list so you don't need to

Reply:
I didn't write that... that bit of code is vendored [directly from pluggy](https://github.com/pytest-dev/pluggy/blob/6e8b6d4e8e0519e50af05b4e4b90b18b924bcbcc/src/pluggy/callers.py#L201-L206) while I [wait for them to consider a PR I submitted](https://github.com/pytest-dev/pluggy/pull/253) that allows you to know _which_ plugin actually returned the result.  (sorry, I thought I had put that in the module docstring, but I guess not!).   So I'm just trying to keep it as similar to their code as possible....
But, they reverse the list because they want wrappers to "sandwich" the call (wrpA, wrpB, hookA, hookB, wrpB, wrpA)

Reply:
@tlambert03 since the module docstring contains some "local" comments, maybe the above comment could also be incorporated. More more more explicit docs! =) Thanks!